src/smartpeak/source/ui/Heatmap2DWidget.cpp

```diff
@@ -6,14 +6,169 @@ namespace SmartPeak
-      if (ImPlot::BeginPlot(plot_title_.c_str(), x_axis_title_.c_str(), y_axis_title_.c_str(), ImVec2(plot_width_ - 25, plot_height_ - 40), imPlotFlags)) {
-        ImPlot::PlotHeatmap(("##" + plot_title_).c_str(), data_.data(), rows_.size(), columns_.size(), data_min_, data_max_, NULL);
-        ImPlot::EndPlot();
+    if (selected_feature_.empty())
+    {
+      ImGui::Text("No features to select");
+    }
+    else
+    {
+      if (ImGui::BeginCombo("Select feature", selected_feature_.c_str()))
+      {
+        for (const auto& feature_name : feature_names)
+        {
+          if (ImGui::Selectable(feature_name.c_str()))
+          {
+            selected_feature_ = feature_name;
+          }
+        }
+        ImGui::EndCombo();
+      }
+
+      // Check if we need to refresh the data
+      Eigen::Tensor<std::string, 1> selected_sample_names = session_handler_.getSelectSampleNamesPlot();
+      Eigen::Tensor<std::string, 1> selected_transitions = session_handler_.getSelectTransitionsPlot();
+      Eigen::Tensor<std::string, 1> selected_transition_groups = session_handler_.getSelectTransitionGroupsPlot();
+      if ((refresh_needed_) || // data changed
+          (selected_feature_ != heatmap_data_.selected_feature_) ||
+          !compareInput(selected_sample_names, heatmap_data_.selected_sample_names_) ||
+          !compareInput(selected_transitions, heatmap_data_.selected_transitions_) ||
+          !compareInput(selected_transition_groups, heatmap_data_.selected_transition_groups_))
+      {
+        session_handler_.getHeatMap(sequence_handler_, heatmap_data_, selected_feature_);
+        // We need to handle "invalid" data - infinite values, or vey high - which will crash ImPlot.
+        invalid_data_ = false;
+        for (int i = 0; i < heatmap_data_.feat_heatmap_row_labels.size(); ++i)
+        {
+          for (int j = 0; j < heatmap_data_.feat_heatmap_col_labels.size(); ++j)
+          {
+            int index = i * heatmap_data_.feat_heatmap_row_labels.size() + j;
+            if (heatmap_data_.feat_heatmap_data.data()[index] > 1e20)
+            {
+              LOGE << selected_feature_ << heatmap_data_.feat_heatmap_col_labels(j) << " / " << heatmap_data_.feat_heatmap_row_labels(i)
+                                        << " contains unexpected high values (" << heatmap_data_.feat_heatmap_data.data()[index] << "), cannot display Heatmap";
+              invalid_data_ = true;
+              break;
+            }
+          }
+        }
+        refresh_needed_ = false;
+      }
+      
+      if (invalid_data_)
+      {
+        ImGui::Text("Some data for this feature seem to be Invalid (unexpected high), cannot display Heatmap");
+      }
+      else
+      {
+        if (heatmap_data_.feat_heatmap_row_labels.size() > 0 && heatmap_data_.feat_heatmap_col_labels.size() > 0)
+        {
+          assert(heatmap_data_.feat_heatmap_data.dimension(0) == heatmap_data_.feat_heatmap_row_labels.size()
```

Comment:
Would it be possible to replace the `assert` with a conditional statement that avoids plotting data instead of stopping the program?  I would also recommend an additional log of the violation.